lesson_10 ready

Gaileks opened the pull request.

Gaileks committed:
lesson_10 ready

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/root@79:
А что это за файл?

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/.idea/kotlinc.xml:
Давай всю папку .idea в .gitignore внесем

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/LearnWordsTrainer.kt:
```diff
@@ -17,7 +19,8 @@ data class Question(
 
 class LearnWordsTrainer(
-    private val wordFileName: String = "words.txt",
+
```
зачем тут перенос

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/.idea/kotlinc.xml:
давай удалим и всю папку .idea в .gitignore.
Удалить желательно в самом intellij idea и закоммитить удаление

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,156 @@
+
+        when (data?.lowercase()) {
+            MAIN_MENU -> {
+                telegramBotService.sendMenu(chatId, json)
```
обрати внимание, json в этом методе всегда передается внутрь telegramBotService, так может можно оставить его внутри telegramBotService и там и создавать?

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,156 @@
+    var lastUpdateId = 0L
+    val json = Json { ignoreUnknownKeys = true }
+    val trainers = HashMap<Long, LearnWordsTrainer>()
     val telegramBotService = TelegramBotService()
```
все таки токен лучше брать из аргументов и передавать в TelegramBotService().
Это даст больше гибкости, например, ты можешь отлаживаться на тестовом токене, а стартовать при публикации на боевом

danelyan replied:
В коде хранить токены точно не стоит. Я бы даже заревочил засвеченный, ведь гит хранит историю.
Так кстати немало людей пострадало из-за утекших amazon aws ключей )

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,156 @@
-                LEARN_WORDS_CLICKED -> {
-                    telegramBotService.sendingQuestionUser(chatId, question)
-                }
+        Thread.sleep(1000)
```
это в константу

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,156 @@
+        if (message?.lowercase() == "/start") {
+            telegramBotService.sendMenu(chatId, json)
+
+        } else if (question != null && question!!.variants.map { it.questionWord }.contains(data?.lowercase())) {
```
обрати внимание, несмотря на question != null, все равно приходится использовать question!!.
Это важный момент.
Так как question мутабельный, в момент между проверкой на null и использованием question переменная может занулиться. Правильнее делать:
question?.let { question ->

}

danelyan reviewed (changes requested):
Пачка правок )

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,156 @@
-                    telegramBotService.sendingQuestionUser(chatId, question)
-                }
+        Thread.sleep(1000)
+        val responseString: String = telegramBotService.getUpdates(lastUpdateId)
```
А вот тут хорошо бы чтобы getUpdates не падал при возникновении сетевых ошибок.
Та ошибка которая у тебя в логах связана с java http-клиентом и с особенностями протокола http2: https://stackoverflow.com/a/67019672.
Не будем углубляться, нам в любом случае нужно учесть ситуацию, когда запрос может выполниться с ошибкой.
Например - сервер телеграм на короткое время стал недоступным или лаг в сети.
Поэтому давай завернем внутри getUpdates(lastUpdateId) запрос в runCatching: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/run-catching.html

Gaileks committed:
Изменил количество слов для выбора при изучении теперь 6 (было 4).
Теперь слово считается выученным после 1 угадывания (было 3).
Изменил расположения кнопок.
Игнорирование ошибок сервера через runCatching.

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,158 @@
-            }
+        waitingTime
+        val responseString: String = runCatching {
+            telegramBotService.getUpdates(lastUpdateId, botToken)
```
я имел в виду брать из аргументов, но передовать в конструктор TelegramBotService() - он же не меняется на протяжении работы приложения.

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,158 @@
+            }
+        if (responseString == "") continue
+
+        val response: Response = json.decodeFromString(responseString)
```
перенеси decodeFromString в метод getUpdates(), туда же перемести json - который конструируется при создании telegramBotService

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,158 @@
+    trainers: HashMap<Long, LearnWordsTrainer>,
+    botToken: String,
+) {
+    val telegramBotService = TelegramBotService()
```
не нужно создавать сервис каждый раз

Gaileks committed:
lesson_10 ready

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/TelegramBotService.kt:
```diff
@@ -13,96 +13,99 @@ const val MAIN_MENU = "/start"
+            val requestBody = SendMessageRequest(
+                chatId,
+                "Выберите перевод слова: ${question.correctAnswer.questionWord}",
+                ReplyMarkup
```
Давай в один столбик сделаем:
<img width="273" alt="Screenshot 2023-11-18 at 16 22 57" src="https://github.com/Gaileks/LearnEnglishWordsTelegramBot/assets/7125622/d6080ac8-e827-403d-8c43-8e77e50efc8e">

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/Telegram.kt:
```diff
@@ -1,75 +1,154 @@
+    val botToken = args[0]
+    var lastUpdateId = 0L
+    val trainers = HashMap<Long, LearnWordsTrainer>()
+    val telegramBotService = TelegramBotService()
```
Давай чтобы вот так было:
```
val telegramBotService = TelegramBotService(botToken = botToken)
```

А из всех методов убрать параметр token

Gaileks committed:
lesson_10 ready

danelyan reviewed (approved):
![yougotthis-martini](https://github.com/Gaileks/LearnEnglishWordsTelegramBot/assets/7125622/3339b61b-9b81-484f-989b-a6e73733be08)


Поздравляю! Ты сделал бота на Kotlin.

Основная часть работы пройдена. Теперь этот бот твой - можешь развивать его как хочешь. Дальше я подготовил тебе развилку из 6 простеньких доработок, все они уже в боте. Однако можешь не ограничиваться только ими, любые идеи приветствуются, просто опиши в личку и мы продумаем детали. 

Дальше как обычно - создаешь Pull Request (даже если что-то еще не работает - шлешь), я пишу что можно улучшить.
